### PR TITLE
Added Training_script with the data loading, transformations, splitti…

### PR DESCRIPTION
**Pull request by SebastianKowalczykiewicz**
…ng. In addition I added a prototype of CNN architecture and a few custiom functions required in training_script

### CODE REVIEW

File: Car_Localizer/Car_Localizer/Training/model_vehicle_classifier.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import os
+os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'
+import torch.nn as nn
+
+
+#CNN Architecture:
+class VehicleClassifier(nn.Module):
+    def __init__(self):
+        super(VehicleClassifier, self).__init__()
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Training/model_vehicle_classifier.py]: you do not need to repeat class name in super() method
super().__init__()

File: Car_Localizer/Car_Localizer/Training/custom_functions.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        channels_sum += torch.mean(data, dim=[0, 2, 3])
+        channels_squared_sum += torch.mean(data ** 2, dim=[0, 2, 3])
+        num_batches += 1
+    # mean: the following function
+    mean = channels_sum / num_batches
+    # std = sqrt(E[X^2] - (E[X])^2)
+    std = (channels_squared_sum / num_batches - mean ** 2) ** 0.5
+    mean = torch.Tensor(mean).tolist()
+    std = torch.Tensor(std).tolist()
+    return mean, std
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Training/custom_functions.py:28]: we should keep everything in PyTorch's tensor manner
so just return the tensor instead a list

```
'''
  Data normalization
  https://www.baeldung.com/cs/normalizing-inputs-artificial-neural-network
'''
def get_mean_std(data:torch.tensor) -> torch.tensor: #get_mean_std 
    assert torch.is_tensor(data), "Example message here..."
    return data.mean(), data.std()
#or
def data_normalization(data:torch.tensor, mean_zero:bool=False) -> torch.tensor: #z_score 
    mu, sigma = data.mean(), data.std()
    return (data - 0.) / sigma if mean_zero else (data - mu) / sigma
```
you can simply use PyTorch's function based

File: Car_Localizer/Car_Localizer/Training/training_script.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+from torchvision import transforms, datasets
+from torch.utils.data import DataLoader, SubsetRandomSampler
+from sklearn.metrics import confusion_matrix
+sns.set_style('darkgrid')
+from model_vehicle_classifier import VehicleClassifier
+from custom_functions import get_mean_and_std, binary_acc
+# from Model import Model_First
+np.random.seed(0)
+torch.manual_seed(0)
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Training/training_script.py:20]: we will use device from `config.py` file, I will add it in next week

File: Car_Localizer/Car_Localizer/Training/training_script.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+print("In our project we're using =>", device)
+root_dir = "/Users/sebastiankowalczykiewicz/Car_Localizer/Car_Localizer/Car_Localizer/Training"
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Training/training_script.py]: `root_dir` should not depend on your path, you need to use local path instead according to python project

File: Car_Localizer/Car_Localizer/Training/training_script.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+print("The data are stored here =>", root_dir)
+
+#Initialize raw training dataset:
+train_dataset_raw = datasets.ImageFolder(root = root_dir + "/train",
+        transform = transforms.Compose([
+        transforms.Resize((64, 64)),
+        transforms.ToTensor()]))
+train_loader_raw = DataLoader(dataset=train_dataset_raw, shuffle=False, batch_size=8)
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Training/training_script.py]: this one is for?

**Review by Siwwyy — CHANGES_REQUESTED** [review]

**Comment by Siwwyy** [comment]
for now @SebastianKowalczykiewicz few stuffs to work on, I will look again if you correct that (and launch your code locally)
but in overall, thanks for good work really! my bows!

File: Car_Localizer/Car_Localizer/Training/model_vehicle_classifier.py (hunk 1 above)
Comment (SebastianKowalczykiewicz) [reply Car_Localizer/Car_Localizer/Training/model_vehicle_classifier.py]: Ok, it will be corrected in the next code release.

File: Car_Localizer/Car_Localizer/Training/training_script.py (hunk 4 above)
Comment (SebastianKowalczykiewicz) [reply Car_Localizer/Car_Localizer/Training/training_script.py]: Corrected

File: Car_Localizer/Car_Localizer/Training/custom_functions.py (hunk 2 above)
Comment (SebastianKowalczykiewicz) [reply Car_Localizer/Car_Localizer/Training/custom_functions.py:28]: I cannot see any other potential processing way how to resolve it properly.
This time I did not want to use custom functions to accomplish it.
#------------------------------------------------------------------------#
Check it pls:
train_loader_raw = DataLoader(dataset=train_dataset_raw, shuffle=False, batch_size=len(train_dataset_raw))
data_raw = next(iter(train_loader_raw))

and after that, data_raw[0].mean() and data_raw[0].std()

Just to clarify:
train_dataset_raw is required to load the inputs and perform the transformations methods like Resize(64,64) as also ToTensor().

File: Car_Localizer/Car_Localizer/Training/training_script.py (hunk 3 above)
Comment (SebastianKowalczykiewicz) [reply Car_Localizer/Car_Localizer/Training/training_script.py:20]: I am going to change the device variable (attached in config.py in near future).
Atm let's keep the current version of var:device dependencies.

File: Car_Localizer/Car_Localizer/Training/training_script.py (hunk 5 above)
Comment (SebastianKowalczykiewicz) [reply Car_Localizer/Car_Localizer/Training/training_script.py]: It has been explained above. 

#Initialize raw training dataset:
train_dataset_raw = datasets.ImageFolder(root = root_dir + "/train",
        transform = transforms.Compose([
        transforms.Resize((64, 64)),
        transforms.ToTensor()]))

train_loader_raw = DataLoader(dataset=train_dataset_raw, shuffle=False, batch_size=len(train_dataset_raw))
data_raw = next(iter(train_loader_raw))

data_raw is used in order to calcualte mean and std on initial version of training dataset after resize and to_tensor method.

File: Car_Localizer/Car_Localizer/Training/training_script.py (hunk 3 above)
Comment (Siwwyy) [reply Car_Localizer/Car_Localizer/Training/training_script.py:20]: lets use config file instead, from config import UsedDevice

File: Car_Localizer/Car_Localizer/Training/custom_functions.py (hunk 2 above)
Comment (Siwwyy) [reply Car_Localizer/Car_Localizer/Training/custom_functions.py:28]: ok, got it

File: Car_Localizer/Car_Localizer/Training/model_vehicle_classifier.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import os
+os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'
+import torch.nn as nn
+
+#CNN Architecture:
+class VehicleClassifier(nn.Module):
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Training/model_vehicle_classifier.py:6]: btw, we have a model already on a branch, but if you like to create new one, please derive from `Model_Base`

**Review by Siwwyy — APPROVED** [review]

**Comment by Siwwyy** [comment]
LGTM in overall!

**Comment by Siwwyy** [comment]
@SebastianKowalczykiewicz any update here?
